### ISSUE TITLE
[V3] We need seperate response and request models

### ISSUE THREAD

**Issue by Stormsys** [justeat/ZendeskApiClient#78]
There are a few issues with sharing models for request and response as we have currently done.

1. Id's are almost never optional from the server, requiring clients to do HasValue and other things to support gets/lists.
2. Updates can be partial models. To solve this we could either setting everything to nullable which conflicts with issue 1. Or have a specific request model options with nullable values.
3. Examples like Requester property on the Ticket model are there for request only making the api unintuitive at times.

@Jetski5822 roping you in for thoughts on the above

My planned action is to separate the models into distinct request and response models.
@kbrimble  how do you feel about this?

**Comment by Stormsys** [comment]
As a second loosely related comment we should consider ditching the abstractions project, it produces a second nuget package that can easily be out of sync with the main one and really the models are pretty tightly coupled with this implementation for most intents and purposes.

**Comment by Jetski5822** [comment]
I agree to ditching the abstractions project, I renamed it from contracts to abstraction to keep what JE had done originally.

One this I like about the existing model, is what you get back is what you use to update. Moving away from that model makes things much let intuitive.

With that said... Things to do with Comment and Requester I totally agree with... The way I see us using these is to keep them on to model, but! change from public to internal, then use the methods on the resources to set them. (this all relies on ditching the abstractions project)

I know the nullable types look cack, but that is the api.

I will try to have a play around on this this weekend, first this first, @Stormsys could you remove the abstractions project and join the two together in a single project?

**Comment by Stormsys** [comment]
@Jetski5822 okay, il do that on a seperate PR (merging the projects together).

Ive been playing around with ideas for models on #79 - but it may not surface as anything.

I know what your saying about reusing models, but its a bit tricky when there is rules and logic applied on models that can come from each "operation" and they vary a lot just on the ticket type. Some options are read only, others required on update and others only required for create this can cause a lot of confusion if its not exlicit.

Things like `new CreateTicketRequest(someTicketReference)` could 'simplify' the api and still allow us to be stricter on data we actually send to the server

**Comment by Jetski5822** [comment]
Saweeet on getting rid of abstractions. I'll play around this weekend and see what I can come up with too.

**Comment by Stormsys** [comment]
@Jetski5822 The deeper i dig into the zendesk documentation the more it makes sense for this to happen(separating the models). as part of this i an surfacing an ErrorResponse model as part of the exception raised on bad requests.

See:
- [Get Response](https://developer.zendesk.com/rest_api/docs/core/tickets#json-format)
- [Create Request](https://developer.zendesk.com/rest_api/docs/core/tickets#request-parameters)
- [Update Request](https://developer.zendesk.com/rest_api/docs/core/tickets#request-body)

I've started work on the idea including adding inline XML doc summaries for each of the parameters.

Im pretty confident this is the right thing to do, Theres little need to reuse models with the partial-update concept `new TicketUpdateRequest(ticketId) { fields to update only changed here }` and it makes the api very clear without having to look at documentation.

see #82 if you wana see where im going with it.
